python/cudf/cudf/core/column/datetime.py

```diff
@@ -517,6 +524,63 @@ def can_cast_safely(self, to_dtype: Dtype) -> bool:
         else:
             return False
 
+    def _with_type_metadata(self, dtype):
```

Comment:
This might need an update to places here to include `DatetimeTZColumn`:

https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/series.py#L1545-L1552
https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/dataframe.py#L3801-L3815
https://github.com/rapidsai/cudf/blob/branch-23.06/python/cudf/cudf/core/dataframe.py#L1702-L1713

Reply:
While we are at it, can we see if can totally remove the special casing in some those places and call `_with_type_metadata` always?

Reply:
Done (and done!)